### PR TITLE
feat!: add whitelist [fixes CHAIN-273 CHAIN-274 CHAIN-275]

### PR DESCRIPTION
**Pull request by elshan-eth**
# Changelog

## DealFactory: 
 - function `deployDeal` has two additional arguments: 
`IConfig.AccessType providersAccessType_` - Access level for a deal  (None, whitelist, blacklist)
`address[] calldata providersAccessList_` - List of addresses in this access list

## Global whitelist 
- add a global whitelist that allows you to limit providers that can create CC
- add two events:
` event WhitelistAccessGranted(address account);`
` event WhitelistAccessRevoked(address account);`

## Deal 
- add a access list and access type (whitelist, blacklist)
- new events: 
` event ProvidersAccessTypeChanged(AccessType newAccessType);`
`event ProviderAddedToAccessList(address provider);`
`event ProviderRemovedFromAccessList(address provider);`

### CODE REVIEW

**Comment by elshan-eth** [comment]
Need to fix ts-client tests

**Comment by AlcibiadesCleinias** [comment]
> Need to fix ts-client tests

`npm test unittests` are resolved successfully 

`npm test integrational` - required created and running infra (currently those tests are not in the CI)

File: src/core/modules/market/Matcher.sol [hunk 1]
```diff
@@ -61,25 +61,35 @@ abstract contract Matcher is Offer, IMatcher {
...
 
         // Go through offers.
         for (uint256 i = 0; i < offers.length; ++i) {
             bytes32 offerId = offers[i];
             Offer memory offer = getOffer(offerId);
+
             if (
-                pricePerWorkerEpoch < offer.minPricePerWorkerEpoch || paymentToken != offer.paymentToken
-                    || !_hasOfferEffectors(offerId, effectors)
+                !deal.hasProviderAccess(offer.provider) || pricePerWorkerEpoch < offer.minPricePerWorkerEpoch
```
Comment (AlcibiadesCleinias) [inline src/core/modules/market/Matcher.sol:79]: deal has provider access - it is really difficult to understand what is happening with these namings

File: src/deal/Config.sol [hunk 2]
```diff
@@ -71,7 +77,9 @@ contract Config is OwnableUpgradableDiamond, IConfig {
         uint256 targetWorkers_,
         uint256 maxWorkersPerProvider_,
         uint256 pricePerWorkerEpoch_,
-        CIDV1[] calldata effectors_
+        CIDV1[] calldata effectors_,
+        AccessType providersAccessType_,
```
Comment (AlcibiadesCleinias) [inline src/deal/Config.sol:81]: why do we have only 1 access type per deal?

Comment (AlcibiadesCleinias) [reply src/deal/Config.sol:81]: I thought we need black list and white list all together

File: src/core/modules/capacity/Whitelist.sol [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    function __Whitelist_init_unchained(bool isWhitelistEnabled) internal onlyInitializing {
+        _getWhitelistStorage().isWhitelistEnabled = isWhitelistEnabled;
+    }
+
+    // #endregion
+
+    // #region ----------------- Public View -----------------
+    function hasAccess(address account) external view returns (bool) {
```
Comment (AlcibiadesCleinias) [inline src/core/modules/capacity/Whitelist.sol:41]: naming is so strange.
mb something:
- `approved`
- `isApproved`
- 'whitelisted'
- 'isWhitelisted'


Btw, For me it seems that WhiteList contract is an abstract only for whitelisting, and if we want to disable those whitelist features in a product we should not do it in the WhiteList itself. We should do it in another abstraction on a layer higher, imho

File: src/core/modules/capacity/Whitelist.sol [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        WhitelistStorage storage s = _getWhitelistStorage();
+        return !s.isWhitelistEnabled || s.isWhitelisted[account];
+    }
+    // #endregion
+
+    // #region ----------------- Public Mutable -----------------
+    function addToWhitelist(address account) external onlyOwner {
```
Comment (AlcibiadesCleinias) [inline src/core/modules/capacity/Whitelist.sol:48]: mb simple `add`?

File: src/core/modules/capacity/Whitelist.sol [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        _getWhitelistStorage().isWhitelisted[account] = true;
+
+        emit AddedToWhitelist(account);
+    }
+
+    function removeFromWhitelist(address account) external onlyOwner {
```
Comment (AlcibiadesCleinias) [inline src/core/modules/capacity/Whitelist.sol:54]: mb simple remove?

File: src/core/modules/market/Matcher.sol [hunk 6]
```diff
@@ -94,29 +104,27 @@ abstract contract Matcher is Offer, IMatcher {
...
                 emit ComputeUnitMatched(peerId, computeUnitId, creationBlock, appCID);
+
+                computeUnitCountInDealByProvider++;
                 freeWorkerSlotsCurrent--;
 
                 if (!isDealMatched) {
                     isDealMatched = true;
                 }
-
-                if (freeWorkerSlotsCurrent == 0) {
```
Comment (AlcibiadesCleinias) [inline src/core/modules/market/Matcher.sol:116]: why did you remove this break?

seems, it is used to stop iteration when we riched target deal worker?

File: src/core/modules/market/Offer.sol [hunk 7]
```diff
@@ -200,7 +200,8 @@ abstract contract Offer is BaseModule, IOffer {
         offerStorage.computeUnitIdsByPeerId[peerId].remove(unitId);
 
         delete offerStorage.computeUnits[unitId];
-        // TODO: event and support in the subgraph.
+
+        emit ComputeUnitRemoved(peerId, unitId);
```
Comment (AlcibiadesCleinias) [inline src/core/modules/market/Offer.sol:204]: I added task for this: https://linear.app/fluence/issue/CHAIN-277/support-cu-removed-in-subgraph-client

**Review by AlcibiadesCleinias — COMMENTED** [review]
About integrational test: we both solve maxWorkersPerProvider TODO in contract in different branches.

That is okey, MB better to merge firstly your PR than I could update my branch and resolve integration test, coz anyway I have it to resolve rather in my branch with my updates or in this, with another updates on maxWorkersPerProvider

File: src/deal/Config.sol (hunk 2 above)
Comment (elshan-eth) [reply src/deal/Config.sol:81]: A whitelist will always be more important than a blacklist. That is, they are not together.

File: src/core/modules/capacity/Whitelist.sol (hunk 3 above)
Comment (elshan-eth) [reply src/core/modules/capacity/Whitelist.sol:41]: I think `hasAccess` is a goode name but `approved` and `isApproved` is better. I will rename it.


I don’t quite understand why this needs to be done at a higher level

File: src/core/modules/market/Matcher.sol (hunk 6 above)
Comment (elshan-eth) [reply src/core/modules/market/Matcher.sol:116]: I moved this rule above

File: test/core/Matcher.t.sol [hunk 8]
```diff
@@ -201,6 +202,27 @@ contract DealMock {
...
+
+    function getComputeUnitCount() external view returns (uint256) {
+        return computeUnitCount;
+    }
+
+    function getComputeUnitCount(address provider) external view returns (uint256) {
+        return computeUnitCountByProvider[provider];
+    }
+
+    function maxWorkersPerProvider() external pure returns (uint256) {
```
Comment (AlcibiadesCleinias) [inline test/core/Matcher.t.sol:225]: why so, why not to save maxWorkerPerProvider into mock?

Comment (AlcibiadesCleinias) [reply test/core/Matcher.t.sol:225]: and simply return through public variable

Comment (elshan-eth) [reply test/core/Matcher.t.sol:225]: I didn't quite get your question

Comment (AlcibiadesCleinias) [reply test/core/Matcher.t.sol:225]: instead of providing method with pure getMax, MB better to leave public variable: https://github.com/fluencelabs/deal/pull/192/files#diff-d75734009f15a4478d68d1a1981ce994c7380b0fa517148aa244488c9fd25df3R177

**Comment by AlcibiadesCleinias** [comment]
> Need to fix ts-client tests

not sure, but mb those fixes could help you: 

- https://github.com/fluencelabs/deal/pull/192/files#diff-904435590130207c656dbea3291d8ce0001307ed4103e804760a97d6d807b434R102
- and this link: https://github.com/fluencelabs/deal/pull/192/files#diff-cc3765daa57de0f2ce6e5a88c6f305afdfcfdf29a891c8ef6df22f27a2385583L107 that leads to implementing maxWorkersPerProvider check on the part of the matcher ts-client

but again, MB better to skip the integrational test in this branch since I fixed it on the part of the feature of this branch: `maxWorkersPerProvider`, and on other part: `deal whitelist` feature I could support on request

File: test/core/Matcher.t.sol (hunk 8 above)
Comment (elshan-eth) [reply test/core/Matcher.t.sol:225]: I think that both ways are normal. I don't see any advantages or disadvantages between them.

File: src/core/modules/market/Matcher.sol (hunk 1 above)
Comment (elshan-eth) [reply src/core/modules/market/Matcher.sol:79]: I change to `isProviderAllowed`

**Review by AlcibiadesCleinias — APPROVED** [review]

**Comment by AlcibiadesCleinias** [comment]
@elshan-eth, let's write a changelog with breaking changes (events, signature of methods) directly in PR and link this changelog in on-chain channel by mentioning: Folex, Masha, Alexey Pr.. Thus, they will be notified that they need to update: miro board, aqua typescript flow, etc.

e.g. for this PR:

# Changelog
- new events in Deal:
```
    event ProvidersAccessTypeChanged(AccessType newAccessType);
    event ProviderAddedToAccessList(address provider);
    event ProviderRemovedFromAccessList(address provider);
```
